# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-04-17T02:37:54.802581907+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ena3rrj/model/ with prefix b8b73363cad248b3
[INFO 2024-04-17T02:37:54.813523368+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:37:54.813791737+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-04-17T02:38:00.624957769+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmx74k0r7/model/ with prefix 7055c1a529fd4a7d
[INFO 2024-04-17T02:38:00.635445378+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2024-04-17T02:38:00.635533744+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmp_omr7_i1 as temporary training directory


[INFO 2024-04-17T02:40:13.99444746+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_omr7_i1/model/ with prefix 9cab4b51079748c2
[INFO 2024-04-17T02:40:14.010161185+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2024-04-17T02:40:14.010231125+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2024-04-17T02:40:14.010274115+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2024-04-17T02:40:15.172110573+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkp07jm_s/model/ with prefix ab35ae2fd93f4111
[INFO 2024-04-17T02:40:15.177189408+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2024-04-17T02:40:17.10373617+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo2nknk6w/model/ with prefix e6278989f90e4af1
[INFO 2024-04-17T02:40:17.127320629+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2024-04-17T02:40:18.37819066+00:00 kernel.cc:1214] Loading model from path /tmp/tmp381tnn48/model/ with prefix fde8a4c94e9b4fe0
[INFO 2024-04-17T02:40:18.383212818+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2024-04-17T02:40:20.571111763+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg79njw14/model/ with prefix 4114a7c6139d4167
[INFO 2024-04-17T02:40:20.599099338+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2024-04-17T02:40:22.740148105+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8_o786uz/model/ with prefix 8a58939b10314844
[INFO 2024-04-17T02:40:22.74611011+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2024-04-17T02:40:24.049243059+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgtycx7yt/model/ with prefix 21c68d7e61f541a8
[INFO 2024-04-17T02:40:24.053159656+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:40:24.053287455+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2024-04-17T02:40:25.419625043+00:00 kernel.cc:1214] Loading model from path /tmp/tmpekki3jis/model/ with prefix e62332b7d2ca4a1c
[INFO 2024-04-17T02:40:25.428216555+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2024-04-17T02:40:27.394828244+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6uvo250d/model/ with prefix 7a2d2771a3d54c34
[INFO 2024-04-17T02:40:27.419371189+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2024-04-17T02:40:30.032154168+00:00 kernel.cc:1214] Loading model from path /tmp/tmpahjduijk/model/ with prefix b1f785d721514d14
[INFO 2024-04-17T02:40:30.043637127+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2024-04-17T02:40:32.490140895+00:00 kernel.cc:1214] Loading model from path /tmp/tmp01hw3l9x/model/ with prefix e5cf84ef345e4311
[INFO 2024-04-17T02:40:32.50957597+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2024-04-17T02:40:33.938303183+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0ozvbs83/model/ with prefix 030f02841a3e498c
[INFO 2024-04-17T02:40:33.944596127+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2024-04-17T02:40:36.130474719+00:00 kernel.cc:1214] Loading model from path /tmp/tmpilr6gypa/model/ with prefix 2119d7d6205f47c4
[INFO 2024-04-17T02:40:36.146295855+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:40:36.146343488+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2024-04-17T02:40:37.477750728+00:00 kernel.cc:1214] Loading model from path /tmp/tmpojy65sl9/model/ with prefix dd5dd46504fb43d6
[INFO 2024-04-17T02:40:37.485464878+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2024-04-17T02:40:39.145047076+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz9i5s09z/model/ with prefix 89d18e919f6a447b
[INFO 2024-04-17T02:40:39.157637459+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2024-04-17T02:40:40.691623931+00:00 kernel.cc:1214] Loading model from path /tmp/tmphg1i6kry/model/ with prefix f7f695f9628a46d9
[INFO 2024-04-17T02:40:40.699049468+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2024-04-17T02:40:42.200430466+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfvd8vhwc/model/ with prefix 241f0000dca6493e
[INFO 2024-04-17T02:40:42.208981165+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2024-04-17T02:40:44.019287179+00:00 kernel.cc:1214] Loading model from path /tmp/tmpak81js1a/model/ with prefix 16ad385d0814488f
[INFO 2024-04-17T02:40:44.033303287+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2024-04-17T02:40:45.648185106+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_0qtejkw/model/ with prefix f48287a979f447e0
[INFO 2024-04-17T02:40:45.662973262+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2024-04-17T02:40:47.268645965+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq3bo6fjf/model/ with prefix 2f5464a8a3904fab
[INFO 2024-04-17T02:40:47.283273685+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:40:47.283338962+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2024-04-17T02:40:49.366216008+00:00 kernel.cc:1214] Loading model from path /tmp/tmprn5qn5am/model/ with prefix e26b6207db634762
[INFO 2024-04-17T02:40:49.386151163+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2024-04-17T02:40:52.357147051+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz4fc92xg/model/ with prefix dd55d1a985c744d8
[INFO 2024-04-17T02:40:52.37381363+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2024-04-17T02:40:53.771331566+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwopea17g/model/ with prefix 3cb4b197c04e4a52
[INFO 2024-04-17T02:40:53.777656162+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2024-04-17T02:40:55.255742915+00:00 kernel.cc:1214] Loading model from path /tmp/tmphi6o_x9b/model/ with prefix ce0e99bf23a240c7
[INFO 2024-04-17T02:40:55.262325784+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2024-04-17T02:40:56.773682087+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbwf0fdde/model/ with prefix 7e62574548f546e5
[INFO 2024-04-17T02:40:56.784363494+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2024-04-17T02:40:58.328542147+00:00 kernel.cc:1214] Loading model from path /tmp/tmpef93cax6/model/ with prefix a2c8a1f237244418
[INFO 2024-04-17T02:40:58.334934244+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:40:58.334979968+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2024-04-17T02:41:00.263378449+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2jrhuhrb/model/ with prefix 2b0af3ad4ed3473a
[INFO 2024-04-17T02:41:00.276214293+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2024-04-17T02:41:01.806551173+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi0kwem8i/model/ with prefix 5e22d874225a428c
[INFO 2024-04-17T02:41:01.817735422+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2024-04-17T02:41:03.263444598+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm2hmtmqw/model/ with prefix 6f878cc122fd4f0b
[INFO 2024-04-17T02:41:03.271309681+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2024-04-17T02:41:04.551415937+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8w9i2rdo/model/ with prefix 3c4666b7f8474eda
[INFO 2024-04-17T02:41:04.557654025+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2024-04-17T02:41:06.240068772+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj2l52_v8/model/ with prefix b4c49503a31c44b9
[INFO 2024-04-17T02:41:06.259712764+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2024-04-17T02:41:08.57450737+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbgldqqe4/model/ with prefix ae67f10cff7241c4
[INFO 2024-04-17T02:41:08.6063268+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:41:08.606453487+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2024-04-17T02:41:10.195856945+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7qtkid5d/model/ with prefix ec991ae3c8d8486d
[INFO 2024-04-17T02:41:10.206850492+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2024-04-17T02:41:13.399926417+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnrrisova/model/ with prefix 6f00e9ec9d6f4f56
[INFO 2024-04-17T02:41:13.40720767+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2024-04-17T02:41:15.157260346+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk4ehtghe/model/ with prefix 5ffa5943ce384279
[INFO 2024-04-17T02:41:15.172381138+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2024-04-17T02:41:16.63326835+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9td1n5u4/model/ with prefix 0b90ee68a1f245fd
[INFO 2024-04-17T02:41:16.643340286+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2024-04-17T02:41:18.271525834+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4nq7z0lm/model/ with prefix 00922176ae8146d6
[INFO 2024-04-17T02:41:18.280515524+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2024-04-17T02:41:20.058277907+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3blbq7x6/model/ with prefix 7bb3976bfff64d46
[INFO 2024-04-17T02:41:20.07917295+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:41:20.07924966+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2024-04-17T02:41:22.90778027+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm3cl5zrg/model/ with prefix 05f9bb9ae5104ffd
[INFO 2024-04-17T02:41:22.917790122+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2024-04-17T02:41:24.728056652+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsylz033d/model/ with prefix f3f2e04f5b934d9a
[INFO 2024-04-17T02:41:24.743132089+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2024-04-17T02:41:26.553915259+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_1u9gisu/model/ with prefix 0795847ba79748f8
[INFO 2024-04-17T02:41:26.570352875+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2024-04-17T02:41:27.97558847+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9c1f01bc/model/ with prefix 2ed3ff23a4fb4ccc
[INFO 2024-04-17T02:41:27.980682333+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2024-04-17T02:41:29.835108249+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn2rk64x1/model/ with prefix 815d519f9c114505
[INFO 2024-04-17T02:41:29.85397354+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2024-04-17T02:41:31.520108531+00:00 kernel.cc:1214] Loading model from path /tmp/tmplna137gp/model/ with prefix ba337114b936494e
[INFO 2024-04-17T02:41:31.530266206+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:41:31.530324581+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2024-04-17T02:41:33.476854928+00:00 kernel.cc:1214] Loading model from path /tmp/tmp81g6rldx/model/ with prefix 2ed3aa87000b451c
[INFO 2024-04-17T02:41:33.496669676+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2024-04-17T02:41:35.047981444+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoc8m02g3/model/ with prefix 43c4619a47854a95
[INFO 2024-04-17T02:41:35.062589393+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2024-04-17T02:41:36.363335164+00:00 kernel.cc:1214] Loading model from path /tmp/tmpso8sme7m/model/ with prefix fdac82475cb54c25
[INFO 2024-04-17T02:41:36.367585178+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2024-04-17T02:41:38.292663602+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuhj163vu/model/ with prefix f5e46d4ea84f47e2
[INFO 2024-04-17T02:41:38.310781873+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2024-04-17T02:41:40.09071402+00:00 kernel.cc:1214] Loading model from path /tmp/tmprjli4_fv/model/ with prefix b821a91765194ad3
[INFO 2024-04-17T02:41:40.107588329+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2024-04-17T02:41:41.681452437+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0yelvtwx/model/ with prefix 832bef3248f44ffe
[INFO 2024-04-17T02:41:41.687020232+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:41:41.687081867+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2024-04-17T02:41:43.714592527+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnzvkn_93/model/ with prefix 102c8a7fe8da457b
[INFO 2024-04-17T02:41:43.722991229+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2024-04-17T02:41:45.74891232+00:00 kernel.cc:1214] Loading model from path /tmp/tmpolj5dd10/model/ with prefix c6190cf46bbb490a
[INFO 2024-04-17T02:41:45.762578747+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2024-04-17T02:41:47.554028806+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqv7hr0dy/model/ with prefix ec26c7d7bd2148e4
[INFO 2024-04-17T02:41:47.571564672+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2024-04-17T02:41:49.059385755+00:00 kernel.cc:1214] Loading model from path /tmp/tmp47923dhd/model/ with prefix b7f2f57ed0354bd8
[INFO 2024-04-17T02:41:49.068573804+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2024-04-17T02:41:50.645545952+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1e4u1h_z/model/ with prefix 54f1671c684e46ac
[INFO 2024-04-17T02:41:50.653687063+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2024-04-17T02:41:51.993594521+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeh2d23xz/model/ with prefix 527f648ff36f46a8
[INFO 2024-04-17T02:41:51.997854086+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:41:51.997978828+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2024-04-17T02:41:54.389324839+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0sn5decs/model/ with prefix 1bf0ace5a72f4c32
[INFO 2024-04-17T02:41:54.407599989+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2024-04-17T02:41:57.006027684+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsffwfvlb/model/ with prefix 1d0102eb3afd4f9d
[INFO 2024-04-17T02:41:57.024644379+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2024-04-17T02:41:58.714479317+00:00 kernel.cc:1214] Loading model from path /tmp/tmpblsn2y5c/model/ with prefix 2b09fba80896475e
[INFO 2024-04-17T02:41:58.72001533+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2024-04-17T02:42:00.717173524+00:00 kernel.cc:1214] Loading model from path /tmp/tmpazry2qp9/model/ with prefix 3ce7ab3edae04cd2
[INFO 2024-04-17T02:42:00.724778309+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2024-04-17T02:42:02.350922015+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmkao559q/model/ with prefix bb5bfecc90b643d3
[INFO 2024-04-17T02:42:02.362378962+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:42:02.362487421+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2024-04-17T02:42:04.01309138+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuikbx5z0/model/ with prefix e145be5191a94b85
[INFO 2024-04-17T02:42:04.02437008+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2024-04-17T02:42:05.336872817+00:00 kernel.cc:1214] Loading model from path /tmp/tmptedi11iy/model/ with prefix 81657a2767a04281
[INFO 2024-04-17T02:42:05.342847081+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2024-04-17T02:42:07.614201194+00:00 kernel.cc:1214] Loading model from path /tmp/tmp57pqxblb/model/ with prefix fc809449a310418f
[INFO 2024-04-17T02:42:07.645655482+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2024-04-17T02:42:09.325875117+00:00 kernel.cc:1214] Loading model from path /tmp/tmp72pj0adq/model/ with prefix 72c8ce4ea85040d1
[INFO 2024-04-17T02:42:09.337005404+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2024-04-17T02:42:11.04077531+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbrf9ywb2/model/ with prefix 1cdd1e54d62e445d
[INFO 2024-04-17T02:42:11.049836189+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2024-04-17T02:42:12.711377903+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj8q5vr6j/model/ with prefix 754b054cc4734552
[INFO 2024-04-17T02:42:12.717716443+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:42:12.717798713+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2024-04-17T02:42:14.173900861+00:00 kernel.cc:1214] Loading model from path /tmp/tmptue8s6av/model/ with prefix e8cc9db30b93469b
[INFO 2024-04-17T02:42:14.182018247+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2024-04-17T02:42:15.992653134+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1pm9vuy_/model/ with prefix 74a24260ecc7402f
[INFO 2024-04-17T02:42:16.013860543+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2024-04-17T02:42:18.917279727+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1q2eyy72/model/ with prefix 405936a2e09a4d42
[INFO 2024-04-17T02:42:18.930442265+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2024-04-17T02:42:20.475201413+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb440_crq/model/ with prefix bf78269e2104422e
[INFO 2024-04-17T02:42:20.484320187+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2024-04-17T02:42:22.110438622+00:00 kernel.cc:1214] Loading model from path /tmp/tmp03zqni4a/model/ with prefix cc243ca8a7ea4bc4
[INFO 2024-04-17T02:42:22.120063265+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2024-04-17T02:42:23.776755001+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkngrwdug/model/ with prefix 85e0c9a0eaf3416a
[INFO 2024-04-17T02:42:23.78583174+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:42:23.785908707+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2024-04-17T02:42:25.659339071+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvakwi_4c/model/ with prefix 1b0336c988394dfc
[INFO 2024-04-17T02:42:25.676700968+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2024-04-17T02:42:27.247231304+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy5f69giu/model/ with prefix 19bf6bd0620e4d0d
[INFO 2024-04-17T02:42:27.254883861+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2024-04-17T02:42:29.454198603+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkqa7ts5u/model/ with prefix 75e4ce467d5e448e
[INFO 2024-04-17T02:42:29.467454338+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2024-04-17T02:42:31.198622638+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf1x0j06x/model/ with prefix 6ed28b3419d44134
[INFO 2024-04-17T02:42:31.208753765+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2024-04-17T02:42:32.578805886+00:00 kernel.cc:1214] Loading model from path /tmp/tmps_mqpo2m/model/ with prefix 78093718e9eb422c
[INFO 2024-04-17T02:42:32.583485009+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2024-04-17T02:42:33.918181113+00:00 kernel.cc:1214] Loading model from path /tmp/tmptc76hlz4/model/ with prefix 191235cb8f3d476f
[INFO 2024-04-17T02:42:33.923128488+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:42:33.923169557+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2024-04-17T02:42:35.422851145+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg7mx43j8/model/ with prefix 3b8fa41718ca4314
[INFO 2024-04-17T02:42:35.430284589+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2024-04-17T02:42:37.306441073+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1dhkbbqh/model/ with prefix ea4d42ddb0a64e30
[INFO 2024-04-17T02:42:37.315690678+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2024-04-17T02:42:38.963627218+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf6nnrgfq/model/ with prefix eab41628773d45e1
[INFO 2024-04-17T02:42:38.975694409+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2024-04-17T02:42:40.825239247+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_3vxk0f0/model/ with prefix 5940f85078444130
[INFO 2024-04-17T02:42:40.838722218+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2024-04-17T02:42:42.653060545+00:00 kernel.cc:1214] Loading model from path /tmp/tmpadaxx2xv/model/ with prefix cf98f268e9bc4566
[INFO 2024-04-17T02:42:42.665718223+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2024-04-17T02:42:44.297057712+00:00 kernel.cc:1214] Loading model from path /tmp/tmpniycj4qq/model/ with prefix 1fdbf36edbac4fd9
[INFO 2024-04-17T02:42:44.308433131+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:42:44.308499731+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2024-04-17T02:42:47.278592187+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj17e5dnv/model/ with prefix 1b7aa9a1f04444a6
[INFO 2024-04-17T02:42:47.299719083+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2024-04-17T02:42:48.860181025+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2xt3vvis/model/ with prefix f42cfa7a73e943a6
[INFO 2024-04-17T02:42:48.867917826+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2024-04-17T02:42:51.080606711+00:00 kernel.cc:1214] Loading model from path /tmp/tmpywpyt6ld/model/ with prefix d5d58cf1a4be4c21
[INFO 2024-04-17T02:42:51.102810262+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2024-04-17T02:42:53.19080937+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_ft227pr/model/ with prefix 01415616dc454c3c
[INFO 2024-04-17T02:42:53.211564532+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2024-04-17T02:42:54.881303332+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjed6eipd/model/ with prefix 40e414e2d0874872
[INFO 2024-04-17T02:42:54.895324551+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:42:54.895388083+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2024-04-17T02:42:56.318071609+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfk757vlt/model/ with prefix ffe50e13babf4e85
[INFO 2024-04-17T02:42:56.32380216+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2024-04-17T02:42:57.853724248+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjebm7ty0/model/ with prefix 90cb0fdcf2d64157
[INFO 2024-04-17T02:42:57.865705727+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2024-04-17T02:42:59.679973028+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2ubyybwc/model/ with prefix 78c665fa8d474209
[INFO 2024-04-17T02:42:59.688264376+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2024-04-17T02:43:01.581643877+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfot3ihxs/model/ with prefix 8401775627394407
[INFO 2024-04-17T02:43:01.60288025+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2024-04-17T02:43:03.174670364+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn4l7os7c/model/ with prefix 0e700b8f8cc44334
[INFO 2024-04-17T02:43:03.187149415+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2024-04-17T02:43:04.613114808+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxprg2n_4/model/ with prefix 8efe7991a7e94307
[INFO 2024-04-17T02:43:04.620289132+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2024-04-17T02:43:06.177499749+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2o1y9qwb/model/ with prefix 004f8b38fb8f4a0c
[INFO 2024-04-17T02:43:06.186672254+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-04-17T02:43:06.186720172+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2024-04-17T02:43:07.800531507+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0gnw_a9d/model/ with prefix aa63a0644baf4724
[INFO 2024-04-17T02:43:07.813759567+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2024-04-17T02:43:09.175846103+00:00 kernel.cc:1214] Loading model from path /tmp/tmpogzym2rd/model/ with prefix 9e9b091bce1e4474
[INFO 2024-04-17T02:43:09.182243899+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2024-04-17T02:43:10.668279836+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsf0uobxe/model/ with prefix 60b7f14cd9c445dd
[INFO 2024-04-17T02:43:10.677474208+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2024-04-17T02:43:12.630900395+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc0ogmtr3/model/ with prefix c673ebf550ae4247
[INFO 2024-04-17T02:43:12.647580543+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 